### ISSUE TITLE
[pickers] AdapterLuxon onChange sends Luxon DateTime instead of js Date

### ISSUE THREAD

**Issue by cyrilchapon** [mui/mui-x#5453]
**Expected**
`onChange` event of DatePicker and others should send a JS Date object.
(just like with DateFns => https://codesandbox.io/s/sqoo8j?file=/demo.tsx)

**Actual**
When using AdapterLuxon, it sends a Luxon DateTime
(see console => https://codesandbox.io/s/materialuipickers-demo-mui-x-forked-73ng8p?file=/demo.tsx)

**Comment by flaviendelangle** [comment]
This is the expected behavior
Why do you think it should pass a JS Date object instead ?

**Comment by cyrilchapon** [comment]
@flaviendelangle, hum, interesting.
After a re-read of documentation, and re-test; it turns out it might just be a doc-clarification need.

My process was the following :

1. Read the general doc on the pickers modules : https://mui.com/x/react-date-pickers/getting-started/
2. Read the paragraphs about date library adapters, etc.
  At that point, I instinctively chose Luxon because I like it, and was pleased to see an official adapter
3. Installed stuff, copy-pasted setup and imports, adapted the LocalizationProvider in my app too use it
  ```
  // or for Luxon
  import { AdapterLuxon } from '@mui/x-date-pickers/AdapterLuxon';

  function App({ children }) {
    return (
      <LocalizationProvider dateAdapter={AdapterDateFns}>
        {children}
      </LocalizationProvider>
    );
  }
  ```
4. Chose a component , `<DatePicker />` in my case.
5. Searched for the documentation of it (https://mui.com/x/react-date-pickers/date-picker/), and understood it
6. Simply copy-pasted the basic example, and started coding with it
  ```
  import * as React from 'react';
  import TextField from '@mui/material/TextField';
  import { DatePicker } from '@mui/x-date-pickers/DatePicker';

  export default function MyPicker() {
    const [value, setValue] = React.useState<Date | null>(null);

    return (
      <DatePicker
        label="Basic example"
        value={value}
        onChange={(newValue) => {
          setValue(newValue);
        }}
        renderInput={(params) => <TextField {...params} />}
      />
    );
  }
  ```
  At this point, my guess was the abstraction layer was actually abstracting the `onChange` and `value` typings, manipulating **internally** with date libraries through @date-io; but simply exposing JS Dates

Indeed, now I'm realizing :
- that all examples are actually using *date-fns*
- which actually doesn't expose a wrapper object (it just manipulates JS Dates)
- that the actual API documentation (https://mui.com/x/api/date-pickers/date-picker/) kinda recalls it, telling `value: any` but `function(value: TValue, keyboardInputValue: string) => void`

Those facts were unclear to me, because :
- The only clue is that hard-to-find `TValue`
- The typing system is so powerful, that once you have copy-pasted
  ```
  const [value, setValue] = React.useState<Date | null>(null);
  ```
  if you start coding
  ```
  <DatePicker
        value={value}
        onChange
  ```
  hovering `onChange` here will show you `(value: Date | null)`, killing your only chance to guess what's happening 😀

**Comment by cyrilchapon** [comment]
To unzoom a little bit, the typings of the `xxxxPicker`s suggest that
> The component will return `onChange` the type you passed inside `value`
while the actual behavior is
> The component will return `onChange` consistent type with the *Adapter* passed to `LocalizationProvider`

If that makes sense.

-------------

My suggestion would be to actually implement typings as it behaves.
Either :
- Type augmentation, to narrow the type based on the adapter ? (I'm adding the adapter, I find in the documentation a mention that I have to type-augment the component accordingly). Material already suggest that at several places in Typescript
- Make the component explicitly generic or parametric, but based on the adapter instead of the `value` ? `<DatePicker adapter={AdapterLuxon} />` or `<DatePicker<typeof AdapterLuxon> />` that would narrow the type of both `value` and `onChange` accordingly

**Comment by flaviendelangle** [comment]
Thanks for your detailed feedback !
Our doc is clearly far from perfect and this issue is something we shoudl actively work on.

> that all examples are actually using date-fns


I proposed to use another adapter than date-fns a few weeks ago.
Date fns do not convert into a different format but manipulate dates. So in the demo, the format returned in indeed a date, which can be confusing.

> that the actual API documentation (https://mui.com/x/api/date-pickers/date-picker/) kinda recalls it, telling value: any

That's probably a limitation of the doc generation.
I'll have a look.

> Make the component explicitly generic or parametric,

You have an explanation of the generic behavior on this PR.
It is not well documented for now.

https://github.com/mui/mui-x/pull/4617

**Comment by flaviendelangle** [comment]
@alexfauquette would you agree to migrate our doc examples to another adapter ?
And if yes which one ?

**Comment by alexfauquette** [comment]
Not a problem to move doc examples. About the choice of the library, I think the best one is `day-js`, because `moment` is not recommended for new projects, and `luxon` does not work out of the box with `maskedInput`

Maybe in [getting started](https://mui.com/x/react-date-pickers/getting-started/) we could add a section "Library usage" between
"Setup" and "Unsupported libraries" to explain that 
- Components interfaces are designed to get either `null` or date object from the chosen library (not a native one).
- It can appear that native date or ISO string can be accepted by components, but we don't support them officially. Which means they cause bugs for some edge cases.

**Comment by flaviendelangle** [comment]
Fixed by #5453
